about.yml generator (depends on shared-config)

yozlet opened the pull request:
Done with @msecret. Based on https://github.com/18F/about_yml . Note that since the `about.yml` file isn't used for much right now, we've skipped questions that seemed like too much work (mainly because they involved the user entering a list of items, e.g. URLs).

Also: this PR is branched from #49 but does not actually use any of the `shared-config` API. (We should probably discuss the best way to share config info.)

yozlet committed:
Add node_modules/ to .gitignore

yozlet committed:
Add JSHint & ESLint configs

yozlet committed:
Initial about.yml generator

cmc333333 commented on .eslintrc.json:
```diff
@@ -0,0 +1,23 @@
+{
```
Not worth holding back this PR, but I think 18F's agreed to the airbnb lint guide.

msecret replied:
Yea we should add it in, I think we just added something temporary for now.

cmc333333 commented on about-yml/index.js:
```diff
@@ -0,0 +1,118 @@
+    type: 'input',
+    name: 'repoName',
+    message: 'What\'s the repo name? (A short name that acts as the project identifier)'
+  },
```
Optimization: We're using the current directory in the [npm generator](https://github.com/18F/18f-cli/blob/210eeff65e368ceb7e9ac9e253a3b6d5f1ed8513/npm/index.js#L23). Do you think that should be a shared prompt? Maybe one that defaults to the current dir?

msecret replied:
Yes definitely. I think in this case, this about-yml should make the prompt, defaulting to the directory and the npm generator should use the value set here (if it is set).

cmc333333 commented on about-yml/index.js:
```diff
@@ -0,0 +1,118 @@
+    type: 'input',
+    name: 'projectFullName',
+    message: 'What\'s the project title? (A few words, title cased, describing the project)'
+  },
```
Optimization: We should probably pull this out and combine the text into a config shared with the [readme](https://github.com/18F/18f-cli/blob/c52e951aca74bf1d57404247a68cba68239f4a58/readme/index.js#L8) generator.

cmc333333 commented on about-yml/index.js:
```diff
@@ -0,0 +1,118 @@
+        } else {
+          return true;
+        }
+      })
```
Should this be simplified to
```js
.filter( prompt => !this.config.get(prompt.name))
```
or is there a nuance I'm missing?

msecret replied:
Yes it should.

cmc333333 commented on about-yml/index.js:
```diff
@@ -0,0 +1,118 @@
+          prompt['default'] = '@TODO';
+        }
+        return prompt;
+      }) 
```
This step feels out of place. I think we should either be explicit (though repetitive) within the prompts definition (maybe we use a constant for "@TODO"?) or this map step should be a `forEach` outside of the `prompting` phase. What do you think?

msecret replied:
I could agree with that, as I just found that theres inputs (like repoName) where we don't want it to be a config, but want @yozlet to weigh in.

cmc333333 commented on about-yml/index.js:
```diff
@@ -0,0 +1,118 @@
+    this.fs.copyTpl(
+      this.templatePath('about.yml'),
+      this.destinationPath('.about.yml'),
+      this.config.getAll()
```
I like the concision!

cmc333333 reviewed (approved):
Added a few nits, but this seems like a very useful step forward.

Commit:
Have npm scaffold use about yml data

As they both share things.

Commit:
Simplify filter conditional function

Commit:
Move shared license string to constants

Because both about yml and npm scaffolds use the license short
name, move it to the shared-config as a constant (at this point,
its not a prompt because the user doesn't have a say, but might be
later).

cmc333333 committed:
Add about-yml to the README.

cmc333333 committed:
Add about-yml to package.json.
